### ISSUE TITLE
Pausing VMs for a while can make them unusable (kernel issues)

### ISSUE THREAD

**Issue by 3hhh** [QubesOS/qubes-issues#5988]
**Qubes OS version**
4.0

**Affected component(s) or functionality**
The VM kernel.
I tested with a debian 10 VM and the 4.19.132-1 kernel.

**Brief summary**
Pausing VMs for a while can make them unusable.

**To Reproduce**
```
qvm-pause vm
#wait for a while, maybe 10-20 minutes
qvm-unpause vm
qvm-run vm xterm
```

**Expected behavior**
`xterm` appears.

**Actual behavior**
Nothing appears. The error message says something about `status code 1`.

**Additional context**
From the log it looks like a kernel time issue to me:

```
[.[0;32m  OK  .[0m] Started .[0;1;39mUpdate time from ClockVM.[0m.
         Starting .[0;1;39mRealtimeKit Scheduling Policy Service.[0m...
[.[0;32m  OK  .[0m] Started .[0;1;39mRealtimeKit Scheduling Policy Service.[0m.
         Starting .[0;1;39mAuthorization Manager.[0m...
[.[0;32m  OK  .[0m] Started .[0;1;39mAuthorization Manager.[0m.
[.[0;32m  OK  .[0m] Started .[0;1;39mDaily man-db regeneration.[0m.

Debian GNU/Linux 10 vm hvc0

[   59.298113] blkfront: xvdi: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
[   78.062653] EXT4-fs (dm-0): mounted filesystem with ordered data mode. Opts: (null)
[ 8420.022228] WARNING: timekeeping: Cycle offset (7894239013655) is larger than allowed by the 'xen' clock's max_cycles value (1981102219259): time overflow danger
[ 8420.022259]          timekeeping: Your kernel is sick, but tries to cope by capping time updates
[ 8420.022456] clocksource: timekeeping watchdog on CPU2: Marking clocksource 'tsc' as unstable because the skew is too large:
[ 8420.022498] clocksource:                       'xen' wd_now: 7df48489754 wd_last: b12f03d930 mask: ffffffffffffffff
[ 8420.022538] clocksource:                       'tsc' cs_now: 13df0e40410c cs_last: 13ed0f0652f mask: ffffffffffffffff
[ 8420.022579] tsc: Marking TSC unstable due to clocksource watchdog
[11847.447178] WARNING: timekeeping: Cycle offset (3247498964369) is larger than allowed by the 'xen' clock's max_cycles value (1981102219259): time overflow danger
[11847.447202]          timekeeping: Your kernel is sick, but tries to cope by capping time updates
[11847.448759] systemd[1]: systemd-journald.service: Main process exited, code=killed, status=6/ABRT
[11847.448847] systemd[1]: systemd-journald.service: Failed with result 'watchdog'.
[11847.449767] systemd[1]: systemd-journald.service: Service has no hold-off time (RestartSec=0), scheduling restart.
[11847.449855] systemd[1]: systemd-journald.service: Scheduled restart job, restart counter is at 1.
[11847.450029] systemd[1]: systemd-journal-flush.service: Succeeded.
[11847.450062] systemd[1]: Stopped Flush Journal to Persistent Storage.
[11847.450104] systemd[1]: Stopping Flush Journal to Persistent Storage...
[11847.450128] systemd[1]: Stopped Journal Service.
[11847.463093] systemd[1]: Starting Journal Service...
[11847.463699] systemd[1]: systemd-logind.service: Main process exited, code=killed, status=6/ABRT
[11847.990514] systemd-journald[1150]: File /run/log/journal/41e4b88828904396880194de46b04943/system.journal corrupted or uncleanly shut down, renaming and replacing.
[11848.023910] u2mfn_release, priv=          (null)

```

**Solutions you've tried**
None.

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
#5987

**Comment by 3hhh** [comment]
Side Note: Sometimes unpausing works. I'd say it's reproducible maybe 50% of the time or so.

**Comment by marmarek** [comment]
I would say it is out of scope. Pausing a VM is a weird thing (for that VM) to do, time normally doesn't jump this way. It isn't surprising at all it has similar weird consequences. If anything, perhaps we should add a warning to the pause button/command that the system inside may not like it.

**Comment by 3hhh** [comment]
> I would say it is out of scope. Pausing a VM is a weird thing (for that VM) to do, time normally doesn't jump this way. It isn't surprising at all it has similar weird consequences. If anything, perhaps we should add a warning to the pause button/command that the system inside may not like it.

While I agree that mentioning potential issues in the manpage for custom kernels, Windows VMs etc. wouldn't hurt, I'd expect pausing & unpausing to work with the mainline Qubes OS kernel.

Afaik Xen pause is identical to Xen suspend with suspend just additionally writing the memory to disk. Since pretty much all companies using Xen migrate VMs between physical machines using the suspend feature, I guess there's some kernel patches for such clock skew issues. Those should ideally be applied.

You might also want some sort of online migration feature for the Qubes Air architecture. ;-)

**Comment by marmarek** [comment]
> Afaik Xen pause is identical to Xen suspend with suspend just additionally writing the memory to disk.

No, it isn't. It just pause VM execution, in-VM kernel is not notified in anyway and for it it looks just like time jump on unpause.

What you've described is VM suspend, which first notify the VM to prepare for it. This means, VM kernel can mitigate issues resulting from this. But it also means (potentially malicious) kernel can simply ignore the request. This means you can't rely on VM suspend for pausing misbehaving VM.

The VM suspend is not callable with user-facing qubes tools.

**Comment by 3hhh** [comment]
> > Afaik Xen pause is identical to Xen suspend with suspend just additionally writing the memory to disk.
> 
> No, it isn't. It just pause VM execution, in-VM kernel is not notified in anyway and for it it looks just like time jump on unpause.
> 
> What you've described is VM suspend, which first notify the VM to prepare for it. This means, VM kernel can mitigate issues resulting from this. But it also means (potentially malicious) kernel can simply ignore the request. This means you can't rely on VM suspend for pausing misbehaving VM.

I see and agree that misbehaving VMs need to be taken into account.

However can't you just give it 30s to suspend gracefully and then use the pause method as fallback if it doesn't comply?
(Similar to `qvm-shutdown` with the kill timeout.)
Or is there some threat that I currently don't see?

Anyway I don't really get the point of the pause feature in Qubes OS at the moment. What good is a pause feature if you have only a 50% chance to unpause?
Is it for memory forensics only?

**Comment by marmarek** [comment]
This is a way to immediately stop a VM from what it is doing - you can access it from domains widget and it is instantaneous - contrary to shutdown/suspend, and also impossible for the VM to stall/ignore. "Kill" would also work for this case, but the implications of miss-click are potentially quite bad (loosing some work) - in case of pausing wrong VM, you can simply unpause it.

**Comment by marmarek** [comment]
Back to your suspend question - what is actually the use case where you want to pause a VM for a long time? You know it doesn't release any resources (besides CPU time), right?

**Comment by 3hhh** [comment]
> This is a way to immediately stop a VM from what it is doing - you can access it from domains widget and it is instantaneous - contrary to shutdown/suspend, and also impossible for the VM to stall/ignore. "Kill" would also work for this case, but the implications of miss-click are potentially quite bad (loosing some work) - in case of pausing wrong VM, you can simply unpause it.
> Note when you pause a VM, domains widget will then list "unpause" and "kill" actions for this VM.

If that's the only use case, you could have employed the following algorithm:
User clicks "kill" -->
1. Pause the VM.
2. Ask for confirmation.
3. a) If Cancel: Unpause.
    b) If Yes: Kill.

> Back to your suspend question - what is actually the use case where you want to pause a VM for a long time? You know it doesn't release any resources (besides CPU time), right?

Yes, but I don't care about RAM - in a few years everyone will have 32GB+ anyway.

What I care about:
- speed of pausing & unpausing >> speed of shutdown & start
- CPU must be off to provide resistance against side channel attacks from potentially malicious VMs

To give you the complete picture:
I don't really like the [design decisions](https://github.com/QubesOS/qubes-issues/issues/832) of [qubes-app-shutdown-idle](https://github.com/QubesOS/qubes-app-shutdown-idle) that resulted in it being in domU - what about the misbehaving VMs refusing to shut down there (while I'm idle)? I wanted to pause/shutdown VMs against potential side channel attacks coming from them.
So I started to develop my own variant with shutdown and pause options in dom0, but hit into this bug which essentially only leaves the shutdown method atm.
(Btw to implement it there's no need to parse any data coming from VMs in dom0 at all.)

**Comment by 3hhh** [comment]
> I wanted to pause/shutdown VMs against potential side channel attacks coming from them.

Admittedly having the memory on disk would also help against on-going side channel attacks (a running VM performing a side channel attack has would have less memory to read from).
So I might try the direct Xen suspend approach as well.

**Comment by andrewdavidwong** [comment]
> I would say it is out of scope. Pausing a VM is a weird thing (for that VM) to do, time normally doesn't jump this way. It isn't surprising at all it has similar weird consequences.

Ok, in that case, I'll close this as "not an issue." If anyone believes this is a mistake, please leave a comment, and we'll be happy to take another look. Thank you.

> If anything, perhaps we should add a warning to the pause button/command that the system inside may not like it.

I agree that we need better UX regarding pausing. I've opened #5993 for this.

**Comment by airelemental** [comment]
> Back to your suspend question - what is actually the use case where you want to pause a VM for a long time?

I pause running VMs during backups (that read from LVM private device), to guarantee consistency. I've also noticed that the larger private volumes which take relatively longer to back up, don't unpause correctly.

Anyway, it's only slightly annoying to have to restart the VM somtime.  I should probably be backing up an LVM snapshot anyway.

I also thought VMs are were paused prior to system suspend, maybe I'm mis-remembering.

**Comment by 3hhh** [comment]
I think this could be fixed by making the kernel more aware of the pause with the following strategy:

To pause:
1. Put the target VM kernel to S3/S4.
2. Wait 0.5s or so.
3. `qvm-pause`

To unpause:
1. `qvm-unpause`
2. Wake the kernel.

I managed to successfully test the pause part via `virsh -c xen:// dompmsuspend [vm] mem` followed by the `qvm-pause`.
`qvm-unpause` also works.

Waking the kernel however failed so far because `virsh` lost the state and doesn't even try (https://github.com/libvirt/libvirt/blob/b94cde18fffe9efe2513e6bb2f281ca2b9756489/src/libxl/libxl_driver.c#L1522) contrary to its manpage (probably a libvirt bug).

S4 sleep to disk inside the VM doesn't seem to work, but that might be some libvirt xml configuration issue. But then I guess S4 would be roughly as slow as stopping & starting a VM. If it's faster, it might of course be awesome as it avoids the memory side channel.

Side note: `dompmsuspend`ed VMs are apparently still scheduled by the hypervisor, i.e. it isn't enough to relaibly make sure it doesn't actually do anything anymore.

**Comment by 3hhh** [comment]
> I think this could be fixed by making the kernel more aware of the pause with the following strategy:
> 
> To pause:
> 
> 1. Put the target VM kernel to S3/S4.
> 2. Wait 0.5s or so.
> 3. `qvm-pause`
> 
> To unpause:
> 
> 1. `qvm-unpause`
> 2. Wake the kernel.
>
> I managed to successfully test the pause part via `virsh -c xen:// dompmsuspend [vm] mem` followed by the `qvm-pause`.
> `qvm-unpause` also works.
> 
> Waking the kernel however failed so far because `virsh` lost the state and doesn't even try (https://github.com/libvirt/libvirt/blob/b94cde18fffe9efe2513e6bb2f281ca2b9756489/src/libxl/libxl_driver.c#L1522) contrary to its manpage (probably a libvirt bug).
> 
> S4 sleep to disk inside the VM doesn't seem to work, but that might be some libvirt xml configuration issue. But then I guess S4 would be roughly as slow as stopping & starting a VM. If it's faster, it might of course be awesome as it avoids the memory side channel.
> 
> Side note: `dompmsuspend`ed VMs are apparently still scheduled by the hypervisor, i.e. it isn't enough to relaibly make sure it doesn't actually do anything anymore.

@marmarek What do you think?

I noticed that you wrote the `libxlDomainPMWakeup` & `libxlDomainPMSuspendForDuration` functions for libvirt a few years back, i.e. I guess you probably know best whether they could be used that way.

An alternative might be `systemctl suspend` inside a VM and then attempt to wake it up, but I didn't get the wakeup done neither so far (are Reset buttons not supported? `xl trigger s3resume` also failed with a strange error).

All of this of course relies on the assumption that the VM kernel may cope better if it's being put to S3/S4 rather than just not scheduled anymore.

**Comment by marmarek** [comment]
The reason for pause being exposed in the GUI at all is to handle not compliant VMs primarily. While some of what you wrote could potentially improve reliability of resuming VM, it will definitely significantly reduce reliability of pausing itself (even by giving a VM notification that it's going to be paused - in such case VM could for example wipe some evidence). What you propose is a different feature. The confusion about its usage will be addressed in https://github.com/QubesOS/qubes-issues/issues/5993, specifically by renaming it to "emergency pause".

You can try `virsh -c xen dompmsuspend ... mem` and then `virsh -c xen dompmwakeup` (not unpause or resume). In many cases it will work, but AFAIR some kernel versions still had issues. But making "pause" using it would actually be a bug, not an improvement.

**Comment by 3hhh** [comment]
> But making "pause" using it would actually be a bug, not an improvement.

Maybe we misunderstood each other there. I didn't suggest it as default, just as an option for a "pause" != "emergency pause" feature.
Something like `qvm-pause --soft` vs `qvm-pause / qvm-pause --hard` with the former being a bit more nice to the target kernel by notifying it and giving it 0.5s to comply.

Anyway as long as I can't test it, I cannot confirm whether it works (qubes-libvirt fails to compile for me with some missing XDR library complaint on fedora 30 & 32 atm).

**Comment by 3hhh** [comment]
For the default Qubes OS kernel I noticed after a bit of experimenting that
1. As it claims, it indeed manages. So no reason to be "nice" to it. 
2. In the process of various service restarts `qubes-gui-agent` appears to fail to cope with the situation.
   (That's why `qvm-run vm xterm` fails, `-p` also mentions that it cannot open a display.)

So a `qvm-run --nogui vm 'systemctl restart qubes-gui-agent'` (or via virsh console) after the unpause fixed all issues for me. 
I guess at least that could be fixed in Qubes OS so that longer pauses are supported with the default Linux kernel.
@andrewdavidwong : So please re-open.

Maybe fix it by calling unpause scripts incl. the `qubes-gui-agent` restart similar to the resume scripts at `/etc/qubes/suspend-post.d/`?
I guess a clock sync on unpause wouldn't hurt either...

Side note:
Since the code of `vm.suspend()` used during Qubes OS suspend is almost identical to the code of `vm.pause()` (cf. https://github.com/QubesOS/qubes-core-admin/blob/master/qubes/vm/qubesvm.py & https://github.com/QubesOS/qubes-core-admin/blob/master/qubes/api/internal.py and notice that both use `libvirt_domain.suspend()`), if no PCI devices are attached, it is highly likely that the Qubes OS suspend on laptop lid close is also affected by this one.
So this issue is likely identical to #4457 (maybe QWT fixes the issue?).
The varius sys-net & sys-usb resume from suspend issues are likely unrelated though as they use the `libvirt_domain.pmSuspendForDuration()`.

**Comment by andrewdavidwong** [comment]
> So a `qvm-run --nogui vm 'systemctl restart qubes-gui-agent'` (or via virsh console) after the unpause fixed all issues for me. 
> I guess at least that could be fixed in Qubes OS so that longer pauses are supported with the default Linux kernel.
> @andrewdavidwong : So please re-open.

@marmarek, what do you think?

**Comment by 3hhh** [comment]
I noticed that my hotfix only works for non-disposable VMs as disposable VMs are shut down when `qubes-gui-agent` is restarted. It also stops the display of existing windows. So it's probably better to fix the underlying root cause (i.e. the X disconnect on unpause).

Here's the relevant `qubes-gui-agent` log for the issue:
```
qubes-gui[578]: qubes_drv disconnected, waiting for possible reconnection
qubes-gui[578]: Waiting on /var/run/xf86-qubes-socket socket...
```

Sometimes (on short pauses) the kernel timekeeping warning doesn't even come up, but the issue and the above log entry does.

The `qvm-run -p vm xterm` output when the issue is present:
```
/usr/lib/qubes/qubes-rpc-multiplexer: 1: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 2: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 3: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 4: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 5: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 6: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 7: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 8: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 9: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 10: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 11: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 12: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 13: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 14: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 15: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 16: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 17: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 18: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 19: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 20: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 21: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 22: /tmp/qubes-session-env: declare: not found
/usr/lib/qubes/qubes-rpc-multiplexer: 23: /tmp/qubes-session-env: declare: not found
xterm: Xt error: Can't open display: :0
```

**Comment by marmarek** [comment]
> The `qvm-run -p vm xterm` output when the issue is present:
> 
> ```
> /usr/lib/qubes/qubes-rpc-multiplexer: 1: /tmp/qubes-session-env: declare: not found
> ```

This looks like totally unrelated, most likely caused by non-default shell.

> So a `qvm-run --nogui vm 'systemctl restart qubes-gui-agent'` (or via virsh console) after the unpause fixed all issues for me.

This restarts X server, so kills all the (GUI) applications. What's the point in pause instead of shutdown then? This is definitely not a "solution" we would add anywhere.

Take a look at logs to find why qubes-gui-agent failed. I would guess it is X server not liking clock jump.

**Comment by 3hhh** [comment]
`Xorg.0.log`:

```
[ 17478.987] (EE) 
Fatal server error:
[ 17478.987] (EE) systemd-logind disappeared (stopped/restarted?)
[ 17478.987] (EE) 
[ 17478.987] (EE) 
Please consult the The X.Org Foundation support 
	 at http://wiki.x.org
 for help. 
[ 17478.987] (EE) Please also check the log file at "/home/user/.local/share/xorg/Xorg.0.log" for additional information.
[ 17478.987] (EE) 
[ 17479.263] (EE) Server terminated with error (1). Closing log file.
```

`.xsession-errors`:

```
(==) Log file: "/home/user/.local/share/xorg/Xorg.0.log", Time: Wed Sep  2 13:16:50 2020
(++) Using config file: "/etc/X11/xorg-qubes.conf"
(==) Using system config directory "/usr/share/X11/xorg.conf.d"
(EE) 
Fatal server error:
(EE) systemd-logind disappeared (stopped/restarted?)
(EE) 
(EE) 
Please consult the The X.Org Foundation support 
         at http://wiki.x.org
 for help. 
(EE) Please also check the log file at "/home/user/.local/share/xorg/Xorg.0.log" for additional information.
(EE) 
(EE) Server terminated with error (1). Closing log file.
/usr/bin/xinit: connection to X server lost
 owned by 0x0
xsettingsd: Took ownership of selection _XSETTINGS_S0
localuser:user being added to access control list
dbus-update-activation-environment: setting QT_ACCESSIBILITY=1
```

So the unpause appears to make the kernel trigger a `systemd-logind` restart which causes the X server to die.

`/tmp/qubes-session-env` errors: I only get these in combination with the issue / dead X / not after `qubes-gui-agent` restart and I use standard bash.

**Comment by 3hhh** [comment]
@andrewdavidwong Not sure what about fixing a dying X server due to a Qubes OS operation is a minor community enhancement.

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong Not sure what about fixing a dying X server due to a Qubes OS operation is a minor community enhancement.

Sure, I'm happy to share my rationale:
 - @marmarek changed the milestone on this to TBD, which indicates that it is not considered a bug. (Every bug is a bug in some specific version of software. If a bug affects multiple versions, we put it on the earliest affected milestone. If it's not tied to any specific version, that suggests that it can't be a bug.)
 - If it's not a bug, it must be either a task or an enhancement (based on how we categorize issues). This seems more like a proposed enhancement.
 - I infer that the Qubes developers consider it to be of minor priority and "help wanted" based on @marmarek's responses here, but it is up to them to assign the priority and "help wanted" status they consider appropriate. (When I assign these, it is simply to try to save them some clerical work by anticipating their needs. I do not consider my own assignments to be as authoritative as theirs are. They're the experts who actually work on the issues -- except in the case of website and doc issues, for example, which may be me.)
 - In general, you should not read too much into the labels here. They can be easily changed as new information or insights come in. This is just a tool to serve the Qubes developers and assist their work (see [reporting bugs and other issues](https://www.qubes-os.org/doc/reporting-bugs/) for more on this). Consider it more like me saying, "Ok, here's my best effort to classify this issue based on what I know right now. Let me know if you have anything better. (Or, if you have permissions, go ahead and change them.)" In some cases, there may be a genuine disagreement based on different values, but in other cases, the disagreement may be based on a misunderstanding or miscommunication. By conferring with each other openly and professionally about these things, we can minimize the latter.

**Comment by 3hhh** [comment]
I see the following chain of events:
1. `qvm-unpause vm`
2. `qubes-sync-time.service / date` in the VM steps the clock in steps larger than 3m to the current time.
3. `systemd` observes a time lapse from the past to the present.
4. Various `systemd` service watchdogs consider their services to be hanging for too long (more than 3m due to the stepping) and restart them. This includes `systemd-logind`.
5. `Xorg` observes `systemd-logind` restarting and decides to shut itself down (https://github.com/systemd/systemd/issues/1163 - apparently this was never fixed on `Xorg` side).
6. `qubes-gui-agent` fails to (re-)connect to the new `Xorg` (requires restart).
7. VM windows close down.
8. If it was a disposable VM, the VM shuts down as no more windows exist.

This leads to various possible fixes:
1. Wait for `Xorg` upstream fix to not crash itself on `systemd-logind` restart (unlikely to happen even though it was fixed in systemd).
2. Make `qubes-gui-daemon` more robust wrt `Xorg` restarts (hard?).
3. Prevent the `Xorg` restart.
4. Previous attempts on pre-warning `systemd`/the kernel e.g. via `S1` sleep (not wanted for at least `qvm-pause`).
5. Smaller time steps in `qubes-sync-time.service` / clock slewing (might have other disadvantages).

The third option seems by far the easiest:

Just remove the line `WatchdogSec=3min` from `/usr/lib/systemd/system/systemd-logind.service` et voilà  - unlimited pauses and unpauses are possible even without notifying the VM beforehand!

Other watchdogs (e.g. `systemd-udevd` and `systemd-journald`) can be disabled as well, but I didn't observe any impact from their service restarts so far. Maybe the udev restart causes e.g. #4658 or #4042 though (suspend and pause work similarly in Qubes OS IIRC).

I suggest to consider that as a default patch for Qubes OS templates - especially since having a working pause/unpause feature is quite cool(tm) for an one-line patch and it likely also helps with suspend/resume. :-)))

**Comment by marmarek** [comment]
Good find @3hhh ! While time jump is not a nice thing for many services, I believe there must be a way to adjust system clock without triggering systemd watchdog. 
Have you verified it is indeed setting the clock that triggers it? `qubes-sync-time.timer` is triggered every 6h, not necessary after every unpause. I'd rather guess systemd treats the pause time as all services hanging and react to it as soon as it can, regardless of system clock.

**Comment by 3hhh** [comment]
By `kernel watchdog` I meant the `kernel clocksource/timekeeping watchdog` of
```
[ 8420.022456] clocksource: timekeeping watchdog on CPU2: Marking clocksource 'tsc' as unstable because the skew is too large:
```
